File: app/src/main/java/com/example/gymapp/data/repository/SmartCoachV2.kt
package com.example.gymapp.data.repository

/** Normative, offline-only Smart Coach v2 inputs shared by Android, iOS and PWA. */
enum class SmartCoachReadinessV2(val wireValue: String) {
    Low("low"),
    Normal("normal"),
    High("high")
}

enum class SmartCoachEquipmentV2(val wireValue: String) {
    Barbell("barbell"),
    Dumbbell("dumbbell"),
    Cable("cable"),
    Machine("machine"),
    Bodyweight("bodyweight"),
    Assisted("assisted"),
    Band("band"),
    Other("other")
}

enum class SmartCoachExerciseFeedbackOutcomeV2(val wireValue: String) {
    OnTarget("onTarget"),
    TooEasy("tooEasy"),
    TooHard("tooHard"),
    EquipmentUnavailable("equipmentUnavailable"),
    TimeCut("timeCut")
}

enum class SmartCoachV2Reason(val wireValue: String) {
    ReadinessRecovery("readinessRecovery"),
    TimeCapped("timeCapped"),
    EquipmentUnavailable("equipmentUnavailable"),
    AvoidedMuscle("avoidedMuscle"),
    ExerciseFeedbackTooHard("exerciseFeedbackTooHard"),
    ExerciseFeedbackRepeatedEasy("exerciseFeedbackRepeatedEasy")
}

enum class SmartCoachExerciseRoleV2(val wireValue: String) {
    Primary("primary"),
    Secondary("secondary"),
    Isolation("isolation"),
    Core("core"),
    Warmup("warmup")
}

data class SmartCoachExerciseFeedbackSignalV2(
    val actualRir: Int? = null,
    val outcome: SmartCoachExerciseFeedbackOutcomeV2 =
        SmartCoachExerciseFeedbackOutcomeV2.OnTarget,
    val ageDays: Int
)

data class SmartCoachContextV2(
    val readiness: SmartCoachReadinessV2? = null,
    val availableMinutes: Int? = null,
    /** Null is unrestricted. A present set must contain 1..8 canonical equipment values. */
    val availableEquipment: Set<SmartCoachEquipmentV2>? = null,
    val musclesToAvoid: Set<String> = emptySet()
) {
    fun isValid(): Boolean =
        (availableMinutes == null || availableMinutes in MIN_AVAILABLE_MINUTES..MAX_AVAILABLE_MINUTES) &&
            (availableEquipment == null || availableEquipment.size in 1..MAX_EQUIPMENT_ITEMS) &&
            musclesToAvoid.size <= MAX_AVOIDED_MUSCLES &&
            musclesToAvoid.all(KNOWN_MUSCLE_IDS::contains)

    companion object {
        const val MIN_AVAILABLE_MINUTES = 30
        const val MAX_AVAILABLE_MINUTES = 120
        const val MAX_EQUIPMENT_ITEMS = 8
        const val MAX_AVOIDED_MUSCLES = 8
        val KNOWN_MUSCLE_IDS: Set<String> = setOf(
            "chest", "shoulders", "biceps", "triceps", "forearms", "abs", "obliques",
            "lats", "upperBack", "lowerBack", "glutes", "quads", "hamstrings",
            "adductors", "calves"
        )
    }
}

data class SmartCoachV2Scenario(
    val requestedEffort: SmartWorkoutEffort,
    val baseSetBudget: Int,
    val role: SmartCoachExerciseRoleV2,
    val equipment: SmartCoachEquipmentV2,
    val primaryMuscles: Set<String> = emptySet(),
    val context: SmartCoachContextV2 = SmartCoachContextV2(),
    /** Newest first. Invalid or stale values are ignored, never coerced. */
    val feedback: List<SmartCoachExerciseFeedbackSignalV2> = emptyList()
)

data class SmartCoachV2Adaptation(
    val eligible: Boolean,
    val appliedEffort: SmartWorkoutEffort,
    val setBudget: Int,
    val targetRir: IntRange,
    val restSeconds: Int,
    val loadAdjustmentSteps: Int,
    val confidenceDelta: Float,
    val freshForSeconds: Int,
    val reasons: List<SmartCoachV2Reason>
)

const val SMART_COACH_V2_CONTRACT_VERSION = 2
const val SMART_COACH_V2_FRESH_FOR_SECONDS = 6 * 60 * 60
const val SMART_COACH_V2_MAX_FEEDBACK_AGE_DAYS = 14
const val SMART_COACH_V2_MAX_FEEDBACK_PER_EXERCISE = 8

/**
 * Small pure policy layer used by the real plan engines and common golden vectors.
 * History-based focus/progression remains in [WorkoutRecommendationEngine].
 */
fun resolveSmartCoachV2(scenario: SmartCoachV2Scenario): SmartCoachV2Adaptation {
    require(scenario.baseSetBudget in 12..24)
    require(scenario.primaryMuscles.all(SmartCoachContextV2.KNOWN_MUSCLE_IDS::contains))
    require(scenario.context.isValid())

    val reasons = mutableListOf<SmartCoachV2Reason>()
    val appliedEffort = when {
        scenario.context.readiness == SmartCoachReadinessV2.Low -> {
            reasons += SmartCoachV2Reason.ReadinessRecovery
            SmartWorkoutEffort.Recovery
        }
        scenario.requestedEffort == SmartWorkoutEffort.Auto -> SmartWorkoutEffort.Standard
        else -> scenario.requestedEffort
    }
    val timeCap = scenario.context.availableMinutes?.let { minutes ->
        (minutes / 3).coerceIn(12, 24)
    }
    val setBudget = minOf(scenario.baseSetBudget, timeCap ?: scenario.baseSetBudget)
    if (setBudget < scenario.baseSetBudget) reasons += SmartCoachV2Reason.TimeCapped

    var eligible = true
    scenario.context.availableEquipment?.let { available ->
        if (scenario.equipment !in available) {
            eligible = false
            reasons += SmartCoachV2Reason.EquipmentUnavailable
        }
    }
    if (scenario.primaryMuscles.any(scenario.context.musclesToAvoid::contains)) {
        eligible = false
        reasons += SmartCoachV2Reason.AvoidedMuscle
    }

    val usableFeedback = scenario.feedback
        .asSequence()
        .take(SMART_COACH_V2_MAX_FEEDBACK_PER_EXERCISE)
        .filter { signal ->
            signal.ageDays in 0..SMART_COACH_V2_MAX_FEEDBACK_AGE_DAYS &&
                (signal.actualRir == null || signal.actualRir in 0..5) &&
                signal.outcome !in setOf(
                    SmartCoachExerciseFeedbackOutcomeV2.EquipmentUnavailable,
                    SmartCoachExerciseFeedbackOutcomeV2.TimeCut
                )
        }
        .toList()
    val latest = usableFeedback.firstOrNull()
    val latestIsHard = latest?.outcome == SmartCoachExerciseFeedbackOutcomeV2.TooHard ||
        latest?.actualRir?.let { it <= 1 } == true
    val firstTwoAreEasy = usableFeedback.take(2).size == 2 &&
        usableFeedback.take(2).all { signal ->
            signal.outcome == SmartCoachExerciseFeedbackOutcomeV2.TooEasy ||
                signal.actualRir?.let { it >= 4 } == true
        }
    val loadAdjustmentSteps = when {
        latestIsHard -> {
            reasons += SmartCoachV2Reason.ExerciseFeedbackTooHard
            -1
        }
        firstTwoAreEasy -> {
            reasons += SmartCoachV2Reason.ExerciseFeedbackRepeatedEasy
            1
        }
        else -> 0
    }
    val targetRir = when {
        latestIsHard || appliedEffort == SmartWorkoutEffort.Recovery -> 3..4
        appliedEffort == SmartWorkoutEffort.Hard &&
            scenario.role in setOf(
                SmartCoachExerciseRoleV2.Primary,
                SmartCoachExerciseRoleV2.Secondary
            ) -> 1..2
        else -> 2..3
    }
    val restSeconds = when (scenario.role) {
        SmartCoachExerciseRoleV2.Primary -> 180
        SmartCoachExerciseRoleV2.Secondary -> 120
        SmartCoachExerciseRoleV2.Isolation -> 75
        SmartCoachExerciseRoleV2.Core,
        SmartCoachExerciseRoleV2.Warmup -> 60
    }
    return SmartCoachV2Adaptation(
        eligible = eligible,
        appliedEffort = appliedEffort,
        setBudget = setBudget,
        targetRir = targetRir,
        restSeconds = restSeconds,
        loadAdjustmentSteps = loadAdjustmentSteps,
        confidenceDelta = if (loadAdjustmentSteps == 0) 0f else 0.05f,
        freshForSeconds = SMART_COACH_V2_FRESH_FOR_SECONDS,
        reasons = reasons.take(4)
    )
}

File: ios/GymApp-iOS/GymApp/Domain/SmartCoachV2.swift
import Foundation

public enum SmartCoachReadinessV2: String, Codable, CaseIterable, Hashable, Sendable {
    case low, normal, high
}

public enum SmartCoachEquipmentV2: String, Codable, CaseIterable, Hashable, Sendable {
    case barbell, dumbbell, cable, machine, bodyweight, assisted, band, other
}

public enum SmartCoachExerciseFeedbackOutcomeV2: String, Codable, CaseIterable, Hashable, Sendable {
    case onTarget, tooEasy, tooHard, equipmentUnavailable, timeCut
}

public enum SmartCoachV2Reason: String, Codable, CaseIterable, Hashable, Sendable {
    case readinessRecovery
    case timeCapped
    case equipmentUnavailable
    case avoidedMuscle
    case exerciseFeedbackTooHard
    case exerciseFeedbackRepeatedEasy
}

public enum SmartCoachExerciseRoleV2: String, Codable, CaseIterable, Hashable, Sendable {
    case primary, secondary, isolation, core, warmup
}

public struct SmartCoachExerciseFeedbackSignalV2: Codable, Hashable, Sendable {
    public let actualRIR: Int?
    public let outcome: SmartCoachExerciseFeedbackOutcomeV2
    public let ageDays: Int

    public init(
        actualRIR: Int? = nil,
        outcome: SmartCoachExerciseFeedbackOutcomeV2 = .onTarget,
        ageDays: Int
    ) {
        self.actualRIR = actualRIR
        self.outcome = outcome
        self.ageDays = ageDays
    }
}

public struct SmartCoachContextV2: Codable, Hashable, Sendable {
    public static let minimumAvailableMinutes = 30
    public static let maximumAvailableMinutes = 120
    public static let maximumEquipmentItems = 8
    public static let maximumAvoidedMuscles = 8
    public static let knownMuscleIDs: Set<String> = [
        "chest", "shoulders", "biceps", "triceps", "forearms", "abs", "obliques",
        "lats", "upperBack", "lowerBack", "glutes", "quads", "hamstrings",
        "adductors", "calves"
    ]

    public let readiness: SmartCoachReadinessV2?
    public let availableMinutes: Int?
    /// Nil is unrestricted. A present set must contain 1...8 canonical values.
    public let availableEquipment: Set<SmartCoachEquipmentV2>?
    public let musclesToAvoid: Set<String>

    public init(
        readiness: SmartCoachReadinessV2? = nil,
        availableMinutes: Int? = nil,
        availableEquipment: Set<SmartCoachEquipmentV2>? = nil,
        musclesToAvoid: Set<String> = []
    ) {
        self.readiness = readiness
        self.availableMinutes = availableMinutes
        self.availableEquipment = availableEquipment
        self.musclesToAvoid = musclesToAvoid
    }

    public var isValid: Bool {
        (availableMinutes.map {
            (Self.minimumAvailableMinutes ... Self.maximumAvailableMinutes).contains($0)
        } ?? true) &&
            (availableEquipment.map { (1 ... Self.maximumEquipmentItems).contains($0.count) } ?? true) &&
            musclesToAvoid.count <= Self.maximumAvoidedMuscles &&
            musclesToAvoid.isSubset(of: Self.knownMuscleIDs)
    }
}

public struct SmartCoachV2Scenario: Sendable {
    public let requestedEffort: SmartWorkoutEffort
    public let baseSetBudget: Int
    public let role: SmartCoachExerciseRoleV2
    public let equipment: SmartCoachEquipmentV2
    public let primaryMuscles: Set<String>
    public let context: SmartCoachContextV2
    /// Newest first. Invalid or stale values are ignored, never coerced.
    public let feedback: [SmartCoachExerciseFeedbackSignalV2]

    public init(
        requestedEffort: SmartWorkoutEffort,
        baseSetBudget: Int,
        role: SmartCoachExerciseRoleV2,
        equipment: SmartCoachEquipmentV2,
        primaryMuscles: Set<String> = [],
        context: SmartCoachContextV2 = SmartCoachContextV2(),
        feedback: [SmartCoachExerciseFeedbackSignalV2] = []
    ) {
        self.requestedEffort = requestedEffort
        self.baseSetBudget = baseSetBudget
        self.role = role
        self.equipment = equipment
        self.primaryMuscles = primaryMuscles
        self.context = context
        self.feedback = feedback
    }
}

public struct SmartCoachV2Adaptation: Equatable, Sendable {
    public let eligible: Bool
    public let appliedEffort: SmartWorkoutEffort
    public let setBudget: Int
    public let targetRIR: ClosedRange<Int>
    public let restSeconds: Int
    public let loadAdjustmentSteps: Int
    public let confidenceDelta: Float
    public let freshForSeconds: Int
    public let reasons: [SmartCoachV2Reason]
}

public let smartCoachV2ContractVersion = 2
public let smartCoachV2FreshForSeconds = 6 * 60 * 60
public let smartCoachV2MaximumFeedbackAgeDays = 14
public let smartCoachV2MaximumFeedbackPerExercise = 8

/// Pure policy layer shared by the real engine and the cross-client golden vectors.
public func resolveSmartCoachV2(_ scenario: SmartCoachV2Scenario) -> SmartCoachV2Adaptation {
    precondition((12 ... 24).contains(scenario.baseSetBudget))
    precondition(scenario.primaryMuscles.isSubset(of: SmartCoachContextV2.knownMuscleIDs))
    precondition(scenario.context.isValid)

    var reasons: [SmartCoachV2Reason] = []
    let appliedEffort: SmartWorkoutEffort
    if scenario.context.readiness == .low {
        appliedEffort = .recovery
        reasons.append(.readinessRecovery)
    } else if scenario.requestedEffort == .auto {
        appliedEffort = .standard
    } else {
        appliedEffort = scenario.requestedEffort
    }
    let timeCap = scenario.context.availableMinutes.map { min(24, max(12, $0 / 3)) }
    let setBudget = min(scenario.baseSetBudget, timeCap ?? scenario.baseSetBudget)
    if setBudget < scenario.baseSetBudget { reasons.append(.timeCapped) }

    var eligible = true
    if let available = scenario.context.availableEquipment,
       !available.contains(scenario.equipment) {
        eligible = false
        reasons.append(.equipmentUnavailable)
    }
    if !scenario.primaryMuscles.isDisjoint(with: scenario.context.musclesToAvoid) {
        eligible = false
        reasons.append(.avoidedMuscle)
    }

    let usableFeedback = scenario.feedback
        .prefix(smartCoachV2MaximumFeedbackPerExercise)
        .filter { signal in
            (0 ... smartCoachV2MaximumFeedbackAgeDays).contains(signal.ageDays) &&
                (signal.actualRIR.map { (0 ... 5).contains($0) } ?? true) &&
                ![.equipmentUnavailable, .timeCut].contains(signal.outcome)
        }
    let latest = usableFeedback.first
    let latestIsHard = latest?.outcome == .tooHard || latest?.actualRIR.map { $0 <= 1 } == true
    let firstTwo = Array(usableFeedback.prefix(2))
    let firstTwoAreEasy = firstTwo.count == 2 && firstTwo.allSatisfy { signal in
        signal.outcome == .tooEasy || signal.actualRIR.map { $0 >= 4 } == true
    }
    let loadAdjustmentSteps: Int
    if latestIsHard {
        loadAdjustmentSteps = -1
        reasons.append(.exerciseFeedbackTooHard)
    } else if firstTwoAreEasy {
        loadAdjustmentSteps = 1
        reasons.append(.exerciseFeedbackRepeatedEasy)
    } else {
        loadAdjustmentSteps = 0
    }

    let targetRIR: ClosedRange<Int>
    if latestIsHard || appliedEffort == .recovery {
        targetRIR = 3 ... 4
    } else if appliedEffort == .hard && [.primary, .secondary].contains(scenario.role) {
        targetRIR = 1 ... 2
    } else {
        targetRIR = 2 ... 3
    }
    let restSeconds = switch scenario.role {
    case .primary: 180
    case .secondary: 120
    case .isolation: 75
    case .core, .warmup: 60
    }
    return SmartCoachV2Adaptation(
        eligible: eligible,
        appliedEffort: appliedEffort,
        setBudget: setBudget,
        targetRIR: targetRIR,
        restSeconds: restSeconds,
        loadAdjustmentSteps: loadAdjustmentSteps,
        confidenceDelta: loadAdjustmentSteps == 0 ? 0 : 0.05,
        freshForSeconds: smartCoachV2FreshForSeconds,
        reasons: Array(reasons.prefix(4))
    )
}
